<a href="https://colab.research.google.com/github/jogabenn/Actividades-MNA-Inteligencia-Artificial-y-Aprendizaje-Automatico/blob/main/Actividades%20Semana%209/Actividad_Semana_9_Equipo50.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Maestría en Inteligencia Artificial Aplicada**
## **Curso: Inteligencia Artificial y Aprendizaje Automático**
### Tecnológico de Monterrey
### Prof. Luis Eduardo Falcón Morales

## **Actividad de la Semana 9**
### **Taxonomía de Métricas de Clasificación**

**Nombres y matrículas de los integrantes del equipo:**

* Jonathan Garza Bennet | A01793038
* Diego Fernando Guerra Burgos | A01793499
* Fernando Maytorena Espinosa de los Monteros | A01110946

---

Esta Tarea se deberá resolver en equipos, de acuerdo a como fueron integrados al inicio del curso.  Como has notado a lo largo del curso, una gran cantidad de métricas se han estado utilizando a lo largo de los diferentes tipos de problemáticas. Aunque en general sabemos que un problema generalmente no se aborda usando una sola métrica, sí conviene tener una idea clara de qué información nos proporciona cada una de las diferentes métricas que se seleccionan. Asimismo tener claro que la comparación de modelos diferentes con métricas diferentes no siempre es del todo válido.<br>
<br>En particular, para los problemas de clasificación se propone una taxonomía de las métricas para clasificación por los autores del siguiente artículo:<br>
<br>C. Ferri, J. Hernández-Orallo, R. Modroiu, An experimental comparison of performance measures for 
classification, Pattern Recognition Letters, Volume 30, Issue 1, 2009, Pages 27-38. 
>https://scholar.google.com/scholar?q=10.1016%2Fj.patrec.2008.08.010%20An%20experimental%20comparison%20of%20performance%20measures%20for%20classification%20Pattern%20recognition%20letters%202009%200167-8655%20FERRI%20C%20HERNANDEZ-ORALLO%2C%20J%20MODROIU%2C%20R%20 

### **Ejercicio 1**<br>
**Realiza un resumen de máximo 1 cuartilla de dicho artículo.**

>El artículo propone una taxonomía de métricas para la clasificación en el aprendizaje automático. La taxonomía se basa en tres dimensiones principales: el tipo de información que proporciona la métrica, el tipo de clasificación para el que está diseñada la métrica y el tipo de entrada que requiere la métrica.<br>
><br>La primera dimensión, el tipo de información, se divide a su vez en tres subdimensiones: exactitud, precisión y recuperación. La exactitud es la proporción de casos correctamente clasificados, la precisión es la proporción de casos positivos correctamente clasificados y la recuperación es la proporción de casos positivos correctamente clasificados.<br>
><br>La segunda dimensión, el tipo de clasificación, se divide a su vez en dos subdimensiones: binaria y multiclase. La clasificación binaria es la clasificación de instancias en dos clases, mientras que la clasificación multiclase es la clasificación de instancias en más de dos clases.<br>
><br>La tercera dimensión, el tipo de entrada, se divide a su vez en dos subdimensiones: etiquetada y sin etiquetar. Los datos etiquetados son aquellos que han sido etiquetados con la etiqueta de clase correcta, mientras que los datos no etiquetados son aquellos que no han sido etiquetados con la etiqueta de clase correcta.<br>
><br>El artículo ofrece una descripción detallada de cada una de las métricas de la taxonomía. Los autores ofrecen ejemplos de cómo se pueden utilizar las métricas en la práctica. También ofrecen un análisis de las ventajas y desventajas de cada una de las métricas.<br>

><br>**Métricas basadas en un umbral y una comprensión cualitativa de error:** precisión, precisión ponderada (aritmética y geométrica), F-measure (F-score) y estadística Kappa: Estas medidas se utilizan cuando queremos que un modelo minimice el número de errores, por lo tanto, se usan comúnmente en aplicaciones de clasificadores. Dentro de esta familia, algunas de estas medidas son más apropiados para conjuntos de datos equilibrados o desequilibrados, para la detección de señales o fallos, o para tareas de recuperación de información.

><br>**Métricas basadas en una comprensión probabilística del error:** es decir, el error medio absoluto (MAE), media absoluta error, el error cuadrático medio (MSE), LogLoss (entropía cruzada), dos versiones de la tasa de probabilidad (rango) y dos medidas para calibración. Estas medidas son especialmente útiles cuando queremos una evaluación de la fiabilidad de los clasificadores, no sólo midiendo cuando fallan, pero si han seleccionado el mal clase con una probabilidad alta o baja. 

><br>**Métricas basadas en qué tan bien el modelo clasifica los ejemplos:** AUC (Flach et al., 2003), que para dos clases equivale a la estadística de Mann-Whitney-Wilcoxon, y está estrechamente relacionado con el concepto de separabilidad. Estos son importantes para muchas aplicaciones, como el diseño de campañas de correo, la gestión de relaciones con el cliente (CRM), sistemas de recomendación, detección de fraude, filtrado de spam, etc., donde se utilizan clasificadores para seleccionar los mejores n instancias de un conjunto de datos o cuando una buena separación de clases es crucial.

><br>Los autores concluyen que la taxonomía de métricas propuesta en el artículo es una herramienta útil para elegir la métrica adecuada para un determinado problema de clasificación.


### **Ejercicio 2**<br>
Con base a la taxonomía de tres tipos de métricas definidas en el artículo (threshold, probabilistic and ranks) analiza los siguientes conjuntos de datos y en cada uno de ellos deberás hacer una propuesta de cuál o cuáles serían las mejores métricas a utilizar en cada caso. Justifica adecuadamente cada propuesta de acuerdo al tipo de problema y tipo de datos.<br><br>
En particular haz énfasis en la métricas que hemos estado estudiando en el curso, las cuales de manera general han sido varias de las principales que usualmente se aplican, o en dado caso algunas de sus variantes. Asimismo determina se lo consideras un problema de clases o no balanceadas. Lee la información de cada liga para que contextuelices el “entendimiento del negocio”.<br><br>
NOTA: Puedes hacer una inspección de los datos, pero no se tienen que ejecutar modelos para su entrenamiento y evaluación.

- **Ejercicio 2.1**<br>
&nbsp;&nbsp;&nbsp;&nbsp; Breast Cancer Wisconsin (Diagnostic) Data Set:<br>
&nbsp;&nbsp;&nbsp;&nbsp; https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)  

- **Ejercicio 2.2**<br>
&nbsp;&nbsp;&nbsp;&nbsp; South German Credit Data Set:<br>
&nbsp;&nbsp;&nbsp;&nbsp; https://archive.ics.uci.edu/ml/datasets/South+German+Credit+%28UPDATE%29

>**Descripción del problema y Objetivo**<br>
>El South German Credit Dataset es una base de datos que contiene información sobre solicitantes de crédito. El conjunto contiene un total de 1,000 observaciones y 21 variables, de las cuales 20 son variables independientes que nos brindan información para predecir la variable "Creditability". La variable independiente nos indica si el sujeto correspondiente a una aplicación de crédito es digno de recibir el crédito o no, y esta variables contiene solamente dos valores: 0: Not Creditworthy (no es digno de crédito), 1: Creditworthy (sí es digno de crédito); por lo tanto este conjunto de datos, y el problema que busca resolver, requieren de un modelo de clasificación.<br><br>
>**Técnicas de Balanceo de Clases**<br>
>Para determinar si se requiere o no aplicar técnicas de balanceo de clases, debemos analizar las observaciones de la variable dependiente. En este conjunto es sencillo determinarlo, pues contamos solamente con dos valores. El conjunto de datos cuenta con 700 solicitudes de crédito de sujetos dignos de recibir un crédito, y 300 solicitudes de crédito de sujetos no dignos de recibir un crédito; visto de otra manera, contamos con 2.33 veces más registros Creditworth, que de registros Not Creditworthy. Si bien una relación 70-30 entre clases no es una disparidad tan elevado, sí es recomendable aplicar técnicas de balanceo de clases; y, considerando que el número total de observaciones del conjunto de datos no es tal elevado, daríamos prioridad a técnicas de *oversampling*, para evitar perder observaciones con técnicas de *undersampling*. Sin embargo, es importante considerar que el *oversampling* puede tender hacia un overfitting a la hora de elegir un algoritmo para desarrollar nuestro modelo de predicción, un buen ejemplo sería Random Forest, el cual tiene como característica evitar el *oversampling*.<br><br>
>**Implicaciones de FN y FP**<br>
>Entendiendo que el objetivo del conjunto de datos es determinar a qué clientes les otorgaremos un crédito, podemos fácilmente interpretar cuáles serían las consecuencias de predecir un falso negativo y un falso positivo. En el caso de predecir un falso negativo, el modelo estaría prácticamente negando la aprobación del crédito a un buen cliente, y por lo tanto, la institución financiera estaría "dejando negocio sobre la mesa". Por el otro lado, el predecir un falso positivo significa otorgarle crédito a un mal pagador, y por lo tanto, aumentar el porcentaje de cartera vencida de la institución financiera. Entonces, para definir el umbral de decisión, debemos determinar qué es más costoso para la institución financiera, si dejar negocio sobre la mesa, o aumentar su porcentaje de cartera vencida. Para definirlo con certeza, tendríamos que analizar el valor que cada cliente deja a la empresa, y el costo que implica cada cliente en cartera vencida; no obstante, sabemos que el valor recibido de un crédito es la ganancia en intereses, el cual suele ser un porcentaje menor al 100% del capital, y por otro lado, el costo de un cliente que se va a cartera vencida implica perder el capital invertido, más el costo de oportunidad de no haber invertido ese capital en un buen cliente que generara ganancias. Por lo tanto, podemos determinar con un alto nivel de certeza que la priordad del modelo debe ser minimizar los falsos positivos.<br><br>
>**Métrica a utilizar**<br>
>Ahora que sabemos que la prioridad de nuestro modelo de clasificación debe ser minimizar el número de falsos positivos, podemos seleccionar una métrica de desempeño que dé priordad a este objetivo, como sería el caso de *Precision*, cuya máxima priordad es precisamente minimizar los falsos positivos.<br><br>$Precision = VP/VP+FP$.

- **Ejercicio 2.3**<br>
&nbsp;&nbsp;&nbsp;&nbsp; Telco customer churn: IBM dataset:<br>
&nbsp;&nbsp;&nbsp;&nbsp; https://www.kaggle.com/datasets/yeanzc/telco-customer-churn-ibm-dataset

- **Ejercicio 2.4**<br>
&nbsp;&nbsp;&nbsp;&nbsp; Qualitative_Bankruptcy Data Set:<br>
&nbsp;&nbsp;&nbsp;&nbsp; https://archive.ics.uci.edu/ml/datasets/Qualitative_Bankruptcy

In [ ]:
# Se incluye aquí todos módulos requeridos.
import pandas as pd
import numpy as np

In [ ]:
#Se cargan los datos del ejercicio.
path = "https://raw.githubusercontent.com/jogabenn/Actividades-MNA-Inteligencia-Artificial-y-Aprendizaje-Automatico/main/Actividades%20Semana%209/Qualitative_Bankruptcy.data.txt"

#Se crea un dataframe pandas sin header, separado por comas sin encabezado
df = pd.read_csv(path, sep=",", header=None)

#Se asigna el nombre de las columnas
df.columns = ['IR','MR','FF','CR','CO','OP','Class']

#Se imprime el encabezado
df.head()

IR MR FF CR CO OP Class
0  P  P  A  A  A  P    NB
1  N  N  A  A  A  N    NB
2  A  A  A  A  A  A    NB
3  P  P  P  P  P  P    NB
4  N  N  P  P  P  N    NB

Se analiza si existen valores faltantes a pesar que se mencione que no existen

In [ ]:
df.isnull().sum()

IR       0
MR       0
FF       0
CR       0
CO       0
OP       0
Class    0
dtype: int64

No existen valores faltantes, ahora se procede a verificar cuántos valores existen de cada clase

In [ ]:
value_counts = df['Class'].value_counts()
value_counts

NB    143
B     107
Name: Class, dtype: int64

Se obtienen los porcentajes de cada clase:

In [ ]:
#Se calculan los porcentajes de cada clase
print('El porcentaje de datos de la clase NB es: {}'.format(round(value_counts[0]/len(df)*100,2))+'%')
print('El porcentaje de datos de la clase B es: {}'.format(round(value_counts[1]/len(df)*100,2))+'%')

El porcentaje de datos de la clase NB es: 57.2%
El porcentaje de datos de la clase B es: 42.8%


Estrictamente no se tiene un ejercicio de clases balanceadas debido a que las muestras no son 50% de cada una. Sin embargo, ambas clases contienen un número suficiente de muestras como para tratarse una primera aproximación como clases balanceadas y estudiar los resultados que se obtengan.

**Descripción del problema y Objetivo**

"Qualitative Bankruptcy database" es un conjunto de datos con métricas cualitativas que califican distintos atributos findancieros de una empresa clasificándolas en 3 categorías:

1. P - Positivo
2. A - Promedio
3. N - Negativo

Los atributos que corresponden a las variables predictoras son los siguientes:

1. Industry risk (IR) : 

  a. Políticas gubernamentales y acuerdos internacionales
  b. Ciclicidad
  c. Grado de competencia
  d. El precio y la estabilidad de la oferta del mercado
  e. El tamaño y el crecimiento de la demanda del mercado
  f. La sensibilidad a los cambios en los factores macroeconómicos
  g. Poder competitivo nacional e internacional
  h. Ciclo de vida del producto.

2. Management risk(MR):

  a. Capacidad y competencia de la dirección
  b. Estabilidad de la gestión
  c. La relación entre dirección/propietario
  d. Administración de recursos humanos
  e. Proceso de crecimiento/rendimiento empresarial
  f. Planificación comercial a corto y largo plazo
  g. Realización y viabilidad

3. Financial Flexibility(FF): 
  a. Financiamiento directo
  b. Financiamiento indirecto
  c. Otra financiación

4. Credibility (CR):  
  a. Historial de crédito
  b. fiabilidad de la información
  c. La relación con las entidades financieras

5. Competitiveness (CO):  
  a. Posición de mercado,
  b. El nivel de las capacidades básicas
  c. Estrategia diferenciada

6. Operating Risk (OP):  
  a. La estabilidad y diversidad de las adquisiciones
  b. La estabilidad de la transacción

Al ser variables categóricas completamente cualitativas, tiene que existir un preprocesamiento previso de todos los indicadores anteriores para poder categorizar en qué rango se encuentra la empresa dentro de todas las opciones disponibles.

La variable de interés con base en las entradas anteriores, resulta en dos estados:

1. B - Bankruptcy
2. NB - Non-Bankruptcy

Como se puede observar, la base de datos consiste en un ejercicio puro de clasificación, ya que no existen ninguna variable numérica continua. Tomando en cuenta la categorización de las métricas, se considera que el conjunto de datos requiere de un tratamiento tipo "threshold", el cual priorice en entedimiento cualitativo del error y nos permita minimizarlo.

El objetivo principal del ejercicio es predecir qué combinación de riesgos resultan en una empresa o entidad que presenta un estado de bancarrota basado en el artículo: *Kim, M.-J., & Han, I. (2003). **The discovery of experts' decision rules from qualitative bankruptcy data using genetic algorithms. Expert Systems with Applications**, 25(4), 637–646. https://doi.org/10.1016/s0957-4174(03)00102-7*

El cual trata por medio de algoritmos genéticos predecir lo anterior, logrando resultados que coinciden con los de los expertos en el tema.

**Técnicas de Balanceo de Clases**

Posterior a análisis de las clases, se tiene que El porcentaje de datos de la clase NB es: 57.2% y porcentaje de datos de la clase B es: 42.8%. Estrictamente no se puede decir que se tenga un conjunto de datos balanceado ya que no se cuenta con ambas clases al 50%, pero se puede decir que se cuenta con un número suficiente de la clase de interés para considerar el ejercicio lo suficientemente balanceado como para requerir técnicas de balanceo. Dependiendo de los resultados de desempeño del clasificador, se puede considerar o no el requerimiento de técnicas de balanceo.

**Implicaciones de FN y FP**

Para este caso, se considera que se deben de minimizar los falsos negativos debido a los efectos negativos para los todos sectores que están involucrados con la bancarrota de una entidad y sus repercusiones sociales. En este caso, el costo de los falsos positivos es menor que el de los falsos negativos y se toma el riesgo de maximizar la métrica de los verdaderos positivos, con el objetivo de capturar el mayor número de los mismos posible.

**Métrica a utilizar**

La métrica de exhaustividad o recall nos permite medir el desempeño sobre qué tantas predicciones positivas se están teniendo en comparación con la suma del total real de positivos, lo que nos permite minimizar la cantidad de falsos negativos, lo cual conincide con lo expresado en el punto anterior.

<br><br>$Recall = VP/(VP+FN$).

---